src/ast/passes/semantic_analyser.cpp

```diff
@@ -1739,9 +1739,12 @@ void SemanticAnalyser::visit(Map &map)
   if (search_val != map_val_.end()) {
     map.type = search_val->second;
   } else {
-    if (is_final_pass()) {
+    // If there is no record of any assignment after the first pass
+    // then it's safe to say this map is undefined
+    if (!is_first_pass()) {
       LOG(ERROR, map.loc, err_) << "Undefined map: " << map.ident;
     }
+    num_unresolved_++;
```

Comment:
Wrap this in a private method? Would make it symmetric with `is_final_pass()`/`is_first_pass()`. Also makes contract a bit more obvious. eg. you cannot decrement a function call.